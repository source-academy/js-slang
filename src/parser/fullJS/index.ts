import { parse } from 'acorn'
import type { Program } from 'estree'

import type { Context } from '../..'
import { FatalSyntaxError } from '../errors'
import type { AcornOptions, Parser } from '../types'
import { positionToSourceLocation } from '../utils'

export class FullJSParser implements Parser<AcornOptions> {
  parse(
    programStr: string,
    context: Context,
    options?: Partial<AcornOptions>,
    throwOnError?: boolean
  ): Program | null {
    try {
      return parse(programStr, {
        sourceType: 'module',
        ecmaVersion: 'latest',
        locations: true,
        ...options
      }) as unknown as Program
    } catch (error) {
      if (error instanceof SyntaxError) {
        error = new FatalSyntaxError(positionToSourceLocation((error as any).loc), error.toString())
      }

      if (throwOnError) throw error
      context.errors.push(error)
    }

    return null
  }

  validate(_ast: Program, _context: Context, _throwOnError: boolean): boolean {
    return true
  }

  toString(): string {
    return 'FullJSParser'
  }
}
