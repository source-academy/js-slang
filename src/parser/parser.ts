/* tslint:disable:max-classes-per-file */
import {
  Options as AcornOptions,
  parse as acornParse,
  parseExpressionAt as acornParseAt,
  Position
} from 'acorn'
import { parse as acornLooseParse } from 'acorn-loose'
import { ancestor, AncestorWalkerFn } from 'acorn-walk/dist/walk'
import * as es from 'estree'
import { Context, ErrorSeverity, ErrorType, Rule, SourceError } from '../types'
import { stripIndent } from '../utils/formatters'
import rules from './rules'
import syntaxBlacklist from './syntaxBlacklist'

export class DisallowedConstructError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public nodeType: string

  constructor(public node: es.Node) {
    this.nodeType = this.formatNodeType(this.node.type)
  }

  get location() {
    return this.node.loc!
  }

  public explain() {
    return `${this.nodeType} are not allowed`
  }

  public elaborate() {
    return stripIndent`
      You are trying to use ${this.nodeType}, which is not allowed (yet).
    `
  }

  /**
   * Converts estree node.type into english
   * e.g. ThisExpression -> 'this' expressions
   *      Property -> Properties
   *      EmptyStatement -> Empty Statements
   */
  private formatNodeType(nodeType: string) {
    switch (nodeType) {
      case 'ThisExpression':
        return "'this' expressions"
      case 'Property':
        return 'Properties'
      default: {
        const words = nodeType.split(/(?=[A-Z])/)
        return words.map((word, i) => (i === 0 ? word : word.toLowerCase())).join(' ') + 's'
      }
    }
  }
}

export class FatalSyntaxError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public constructor(public location: es.SourceLocation, public message: string) {}

  public explain() {
    return this.message
  }

  public elaborate() {
    return 'There is a syntax error in your program'
  }
}

export class MissingSemicolonError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public constructor(public location: es.SourceLocation) {}

  public explain() {
    return 'Missing semicolon at the end of statement'
  }

  public elaborate() {
    return 'Every statement must be terminated by a semicolon.'
  }
}

export class TrailingCommaError implements SourceError {
  public type: ErrorType.SYNTAX
  public severity: ErrorSeverity.WARNING
  public constructor(public location: es.SourceLocation) {}

  public explain() {
    return 'Trailing comma'
  }

  public elaborate() {
    return 'Please remove the trailing comma'
  }
}

export function parseAt(source: string, num: number) {
  let theNode: acorn.Node | undefined
  try {
    theNode = acornParseAt(source, num)
  } catch (error) {
    return undefined
  }
  return theNode
}

export function parse(source: string, context: Context, fallbackToLooseParse: boolean = false) {
  let program: es.Program | undefined
  try {
    program = (acornParse(source, createAcornParserOptions(context)) as unknown) as es.Program
    ancestor(program as es.Node, walkers, undefined, context)
  } catch (error) {
    if (error instanceof SyntaxError) {
      // tslint:disable-next-line:no-any
      const loc = (error as any).loc
      const location = {
        start: { line: loc.line, column: loc.column },
        end: { line: loc.line, column: loc.column + 1 }
      }
      context.errors.push(new FatalSyntaxError(location, error.toString()))
    } else {
      throw error
    }
  }
  const hasErrors = context.errors.find(m => m.severity === ErrorSeverity.ERROR)
  if (program && !hasErrors) {
    return program
  } else if (fallbackToLooseParse) {
    return looseParse(source, context)
  } else {
    return undefined
  }
}

const createAcornParserOptions = (context: Context): AcornOptions => ({
  sourceType: 'module',
  ecmaVersion: 6,
  locations: true,
  // tslint:disable-next-line:no-any
  onInsertedSemicolon(end: any, loc: any) {
    context.errors.push(
      new MissingSemicolonError({
        end: { line: loc.line, column: loc.column + 1 },
        start: loc
      })
    )
  },
  // tslint:disable-next-line:no-any
  onTrailingComma(end: any, loc: Position) {
    context.errors.push(
      new TrailingCommaError({
        end: { line: loc.line, column: loc.column + 1 },
        start: loc
      })
    )
  }
})

// Names-extractor needs comments
export function parseForNames(source: string): [es.Program, acorn.Comment[]] {
  let comments: acorn.Comment[] = []
  const options: AcornOptions = {
    sourceType: 'module',
    ecmaVersion: 6,
    locations: true,
    onComment: comments
  }
  let program: es.Program | undefined
  try {
    program = (acornParse(source, options) as unknown) as es.Program
  } catch {
    comments = []
    program = acornLooseParse(source, options)
  }

  return [program, comments]
}

export function looseParse(source: string, context: Context) {
  const program = (acornLooseParse(
    source,
    createAcornParserOptions(context)
  ) as unknown) as es.Program
  return program
}

function createWalkers(
  allowedSyntaxes: { [nodeName: string]: number },
  parserRules: Rule<es.Node>[]
) {
  const newWalkers = new Map<string, AncestorWalkerFn<Context>>()
  const visitedNodes = new Set<es.Node>()

  // Provide callbacks checking for disallowed syntaxes, such as case, switch...
  const syntaxPairs = Object.entries(allowedSyntaxes)
  syntaxPairs.map(pair => {
    const syntax = pair[0]
    newWalkers.set(syntax, (node: es.Node, context: Context, ancestors: [es.Node]) => {
      if (!visitedNodes.has(node)) {
        visitedNodes.add(node)

        if (context.chapter < allowedSyntaxes[node.type]) {
          context.errors.push(new DisallowedConstructError(node))
        }
      }
    })
  })

  // Provide callbacks checking for rule violations, e.g. no block arrow funcs, non-empty lists...
  parserRules.forEach(rule => {
    const checkers = rule.checkers
    const syntaxCheckerPair = Object.entries(checkers)
    syntaxCheckerPair.forEach(pair => {
      const syntax = pair[0]
      const checker = pair[1]
      const oldCheck = newWalkers.get(syntax)!
      const newCheck = (node: es.Node, context: Context, ancestors: es.Node[]) => {
        if (typeof rule.disableOn !== 'undefined' && context.chapter >= rule.disableOn) {
          return
        }
        const errors = checker(node, ancestors)
        errors.forEach(e => context.errors.push(e))
      }
      newWalkers.set(syntax, (node: es.Node, context: Context<any>, ancestors) => {
        oldCheck(node, context, ancestors)
        newCheck(node, context, ancestors)
      })
    })
  })

  return mapToObj(newWalkers)
}

const mapToObj = (map: Map<string, any>) =>
  Array.from(map).reduce((obj, [k, v]) => Object.assign(obj, { [k]: v }), {})

const walkers: { [name: string]: AncestorWalkerFn<Context> } = createWalkers(syntaxBlacklist, rules)
