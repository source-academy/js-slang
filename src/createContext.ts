// Variable determining chapter of Source is contained in this file.

import { GLOBAL, GLOBAL_KEY_TO_ACCESS_NATIVE_STORAGE } from './constants'
import { AsyncScheduler } from './schedulers'
import * as list from './stdlib/list'
import { list_to_vector } from './stdlib/list'
import { listPrelude } from './stdlib/list.prelude'
import { nonDetPrelude } from './stdlib/non-det.prelude'
import * as misc from './stdlib/misc'
import * as parser from './stdlib/parser'
import * as stream from './stdlib/stream'
import { streamPrelude } from './stdlib/stream.prelude'
import { Context, CustomBuiltIns, Value, Variant } from './types'
import * as operators from './utils/operators'
import * as gpu_lib from './gpu/lib'
import { stringify } from './utils/stringify'
import { lazyListPrelude } from './stdlib/lazyList.prelude'
export class LazyBuiltIn {
  func: (...arg0: any) => any
  evaluateArgs: boolean
  constructor(func: (...arg0: any) => any, evaluateArgs: boolean) {
    this.func = func
    this.evaluateArgs = evaluateArgs
  }
}
const createEmptyRuntime = () => ({
  break: false,
  debuggerOn: true,
  isRunning: false,
  environments: [],
  value: undefined,
  nodes: []
})

const createEmptyDebugger = () => ({
  observers: { callbacks: Array<() => void>() },
  status: false,
  state: {
    it: (function*(): any {
      return
    })(),
    scheduler: new AsyncScheduler()
  }
})

const createGlobalEnvironment = () => ({
  tail: null,
  name: 'global',
  head: {}
})

export const createEmptyContext = <T>(
  chapter: number,
  variant: Variant = 'default',
  externalSymbols: string[],
  externalContext?: T
): Context<T> => {
  if (!Array.isArray(GLOBAL[GLOBAL_KEY_TO_ACCESS_NATIVE_STORAGE])) {
    GLOBAL[GLOBAL_KEY_TO_ACCESS_NATIVE_STORAGE] = []
  }
  const length = GLOBAL[GLOBAL_KEY_TO_ACCESS_NATIVE_STORAGE].push({
    globals: { variables: new Map(), previousScope: null },
    operators: new Map(Object.entries(operators)),
    gpu: new Map(Object.entries(gpu_lib))
  })
  return {
    chapter,
    externalSymbols,
    errors: [],
    externalContext,
    runtime: createEmptyRuntime(),
    numberOfOuterEnvironments: 1,
    prelude: null,
    debugger: createEmptyDebugger(),
    contextId: length - 1,
    executionMethod: 'auto',
    variant
  }
}

export const ensureGlobalEnvironmentExist = (context: Context) => {
  if (!context.runtime) {
    context.runtime = createEmptyRuntime()
  }
  if (!context.runtime.environments) {
    context.runtime.environments = []
  }
  if (context.runtime.environments.length === 0) {
    context.runtime.environments.push(createGlobalEnvironment())
  }
}

const defineSymbol = (context: Context, name: string, value: Value) => {
  const globalEnvironment = context.runtime.environments[0]
  Object.defineProperty(globalEnvironment.head, name, {
    value,
    writable: false,
    enumerable: true
  })
  GLOBAL[GLOBAL_KEY_TO_ACCESS_NATIVE_STORAGE][context.contextId].globals.variables.set(name, {
    kind: 'const',
    getValue: () => value
  })
}

// Defines a builtin in the given context
// If the builtin is a function, wrap it such that its toString hides the implementation
export const defineBuiltin = (context: Context, name: string, value: Value) => {
  if (typeof value === 'function') {
    const wrapped = (...args: any) => value(...args)
    const funName = name.split('(')[0].trim()
    const repr = `function ${name} {\n\t[implementation hidden]\n}`
    wrapped.toString = () => repr

    defineSymbol(context, funName, wrapped)
  } else if (value instanceof LazyBuiltIn) {
    const wrapped = (...args: any) => value.func(...args)
    const funName = name.split('(')[0].trim()
    const repr = `function ${name} {\n\t[implementation hidden]\n}`
    wrapped.toString = () => repr
    defineSymbol(context, funName, new LazyBuiltIn(wrapped, value.evaluateArgs))
  } else {
    defineSymbol(context, name, value)
  }
}

export const importExternalSymbols = (context: Context, externalSymbols: string[]) => {
  ensureGlobalEnvironmentExist(context)

  externalSymbols.forEach(symbol => {
    defineSymbol(context, symbol, GLOBAL[symbol])
  })
}

/**
 * Imports builtins from standard and external libraries.
 */
export const importBuiltins = (context: Context, externalBuiltIns: CustomBuiltIns) => {
  ensureGlobalEnvironmentExist(context)

  const rawDisplay = (v: Value, s: string) =>
    externalBuiltIns.rawDisplay(v, s, context.externalContext)
  const display = (v: Value, s: string) => (rawDisplay(stringify(v), s), v)
  const prompt = (v: Value) => externalBuiltIns.prompt(v, '', context.externalContext)
  const alert = (v: Value) => externalBuiltIns.alert(v, '', context.externalContext)
  const visualiseList = (v: Value) => externalBuiltIns.visualiseList(v, context.externalContext)

  if (context.chapter >= 1) {
    defineBuiltin(context, 'runtime()', misc.runtime)
    defineBuiltin(context, 'display(val)', display)
    defineBuiltin(context, 'raw_display(str)', rawDisplay)
    defineBuiltin(context, 'stringify(val)', stringify)
    defineBuiltin(context, 'error(str)', misc.error_message)
    defineBuiltin(context, 'prompt(str)', prompt)
    defineBuiltin(context, 'is_number(val)', misc.is_number)
    defineBuiltin(context, 'is_string(val)', misc.is_string)
    defineBuiltin(context, 'is_function(val)', misc.is_function)
    defineBuiltin(context, 'is_boolean(val)', misc.is_boolean)
    defineBuiltin(context, 'is_undefined(val)', misc.is_undefined)
    defineBuiltin(context, 'parse_int(str, radix)', misc.parse_int)
    defineBuiltin(context, 'undefined', undefined)
    defineBuiltin(context, 'NaN', NaN)
    defineBuiltin(context, 'Infinity', Infinity)
    // Define all Math libraries
    const props = Object.getOwnPropertyNames(Math)
    for (const prop of props) {
      defineBuiltin(context, 'math_' + prop, Math[prop])
    }
  }

  if (context.chapter >= 2) {
    // List library

    if (context.variant === 'lazy') {
      defineBuiltin(context, 'pair(left, right)', new LazyBuiltIn(list.pair, false))
      defineBuiltin(context, 'list(...values)', new LazyBuiltIn(list.list, false))
      defineBuiltin(context, 'is_pair(val)', new LazyBuiltIn(list.is_pair, true))
      defineBuiltin(context, 'head(xs)', new LazyBuiltIn(list.head, true))
      defineBuiltin(context, 'tail(xs)', new LazyBuiltIn(list.tail, true))
      defineBuiltin(context, 'is_null(val)', new LazyBuiltIn(list.is_null, true))
      defineBuiltin(context, 'draw_data(xs)', new LazyBuiltIn(visualiseList, true))
    } else {
      defineBuiltin(context, 'pair(left, right)', list.pair)
      defineBuiltin(context, 'is_pair(val)', list.is_pair)
      defineBuiltin(context, 'head(xs)', list.head)
      defineBuiltin(context, 'tail(xs)', list.tail)
      defineBuiltin(context, 'is_null(val)', list.is_null)
      defineBuiltin(context, 'list(...values)', list.list)
      defineBuiltin(context, 'draw_data(xs)', visualiseList)
    }
  }

  if (context.chapter >= 3) {
    defineBuiltin(context, 'set_head(xs, val)', list.set_head)
    defineBuiltin(context, 'set_tail(xs, val)', list.set_tail)
    defineBuiltin(context, 'array_length(arr)', misc.array_length)
    defineBuiltin(context, 'is_array(val)', misc.is_array)

    // Stream library
    defineBuiltin(context, 'stream_tail(stream)', stream.stream_tail)
    defineBuiltin(context, 'stream(...values)', stream.stream)
    defineBuiltin(context, 'list_to_stream(xs)', stream.list_to_stream)
  }

  if (context.chapter >= 4) {
    defineBuiltin(context, 'parse(program_string)', (str: string) =>
      parser.parse(str, createContext(context.chapter))
    )
    defineBuiltin(
      context,
      'apply_in_underlying_javascript(fun, args)',
      // tslint:disable-next-line:ban-types
      (fun: Function, args: Value) => fun.apply(fun, list_to_vector(args))
    )
  }

  if (context.chapter >= 100) {
    defineBuiltin(context, 'is_object(val)', misc.is_object)
    defineBuiltin(context, 'is_NaN(val)', misc.is_NaN)
    defineBuiltin(context, 'has_own_property(obj, prop)', misc.has_own_property)
    defineBuiltin(context, 'alert(val)', alert)
    // tslint:disable-next-line:ban-types
    defineBuiltin(context, 'timed(fun)', (f: Function) =>
      misc.timed(context, f, context.externalContext, externalBuiltIns.rawDisplay)
    )
  }
}

function importPrelude(context: Context) {
  let prelude = ''
  if (context.chapter >= 2) {
    prelude += context.variant === 'lazy' ? lazyListPrelude : listPrelude
  }
  if (context.chapter >= 3) {
    prelude += streamPrelude
  }

  if (context.variant === 'non-det') {
    prelude += nonDetPrelude
  }

  if (prelude !== '') {
    context.prelude = prelude
  }
}

const defaultBuiltIns: CustomBuiltIns = {
  rawDisplay: misc.rawDisplay,
  // See issue #5
  prompt: misc.rawDisplay,
  // See issue #11
  alert: misc.rawDisplay,
  visualiseList: (v: Value) => {
    throw new Error('List visualizer is not enabled')
  }
}

const createContext = <T>(
  chapter = 1,
  variant: Variant = 'default',
  externalSymbols: string[] = [],
  externalContext?: T,
  externalBuiltIns: CustomBuiltIns = defaultBuiltIns
) => {
  const context = createEmptyContext(chapter, variant, externalSymbols, externalContext)

  importBuiltins(context, externalBuiltIns)
  importPrelude(context)
  importExternalSymbols(context, externalSymbols)

  return context
}

export default createContext
