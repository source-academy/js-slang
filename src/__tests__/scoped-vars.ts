import * as es from 'estree'
import { default as createContext } from '../createContext'
import { looseParse, parse } from '../parser/parser'
import { getAllOccurrencesInScope, lookupDefinition, scopeVariables } from '../scoped-vars'
/* tslint:disable:max-classes-per-file */

/**
 * Helper function to get the parsed estree program to test the function
 */
function testParseHelper(
  code: string,
  // default to 4
  chapter: number = 4,
  useLooseParser: boolean = false
): es.Program | undefined {
  if (useLooseParser) {
    return looseParse(code, createContext(chapter))
  } else {
    return parse(code, createContext(chapter))
  }
}

class Target {
  public name: string
  public row: number
  public col: number

  constructor(name: string, row: number, col: number) {
    this.name = name
    this.row = row
    this.col = col
  }
}

class CodeTest {
  public label: string
  public _code: string
  public targets: Target[]
  public get code(): string {
    return this._code
      .trim()
      .split('\n')
      .map(line => line.trim())
      .join('\n')
  }
  public testAndAdd: (testFn: (code: string) => void) => void

  constructor(label: string, code: string, targets: Target[]) {
    this.label = label
    this._code = code
    this.targets = targets
  }
}

// Makes a more readable debuggable result
function result(testCase: CodeTest, target: Target, value: any) {
  return {
    label: testCase.label,
    code: testCase.code,
    target,
    value
  }
}

const variableDefinitionTests = [
  new CodeTest(
    'Test 1',
    `
      let target = 1;
      target = 2;
      {
        const target = 3;
        target + 4;
      }
    `,
    [new Target('target', 5, 3)]
  ),
  new CodeTest(
    'Test 2',
    `
      let target = 1;
      target = 2;
      target + 99;
    `,
    [new Target('target', 3, 3), new Target('MISSING', 3, 3)]
  )
]

const functionScopeTests = [
  new CodeTest(
    'Test 1',
    `
      let target = 1;
      target = 2;
      function test(target) {
        const x = 3;
        target + 4;
        function y() {
          target * 99;
        }
      }
    `,
    [new Target('target', 7, 7), new Target('target', 5, 6)]
  )
]

const arrowFunctionScopeTests = [
  new CodeTest(
    'Test 1',
    `
      let target = 1;
      function test(target) {
        const arrowFn = target => target + 1;
        const x = target * 777777;
        const nestedArrowFn = (y) => (target) => 1 + target;
      }
    `,
    [new Target('target', 5, 53)]
  )
]

const conditionalsLoopsTests = [
  new CodeTest(
    'Test 1',
    `
      let target = 1;
      if (target < 2) {
        target = 2 + 3;
      } else if (target > 2) {
        target = "nooooooo";
      } else {
        target = 99999;
      }
    `,
    [new Target('target', 5, 7)]
  ),
  new CodeTest(
    'Test 2',
    `
      let target = 2;
      for (let i = 0; i < 10; i++) {
        target += 99;
      }
    `,
    [new Target('target', 3, 5), new Target('target', 2, 13), new Target('i', 5, 2)]
  ),
  new CodeTest(
    'Test 3',
    `
      let target = 2;
      while (target > 0) {
        const x = "i am coronavirus";
        target -= 0.1;
      }
    `,
    [new Target('target', 4, 4)]
  )
]

test('Going to definitions of ordinary variable definitions', () => {
  const actuals: any = []
  variableDefinitionTests.forEach(testCase => {
    const program = testParseHelper(testCase.code)
    if (program === undefined) {
      return
    }
    const scopedTree = scopeVariables(program)
    testCase.targets.forEach(target => {
      actuals.push(
        result(testCase, target, lookupDefinition(target.name, target.row, target.col, scopedTree))
      )
    })
  })
  expect(actuals).toMatchSnapshot()
})

test('Scoped based refactoring of ordinary variable definitions', () => {
  const actuals: any = []
  variableDefinitionTests.forEach(testCase => {
    const scopedTree = testParseHelper(testCase.code)
    if (scopedTree === undefined) {
      return
    }
    testCase.targets.forEach(target => {
      actuals.push(
        result(
          testCase,
          target,
          getAllOccurrencesInScope(target.name, target.row, target.col, scopedTree)
        )
      )
    })
  })
  expect(actuals).toMatchSnapshot()
})

test('Going to definitions of with function scopes', () => {
  const actuals: any = []
  functionScopeTests.forEach(testCase => {
    const program = testParseHelper(testCase.code)
    if (program === undefined) {
      return
    }
    const scopedTree = scopeVariables(program)
    testCase.targets.forEach(target => {
      actuals.push(
        result(testCase, target, lookupDefinition(target.name, target.row, target.col, scopedTree))
      )
    })
  })
  expect(actuals).toMatchSnapshot()
})

test('Scoped based refactoring with function scopes', () => {
  const actuals: any = []
  functionScopeTests.forEach(testCase => {
    const scopedTree = testParseHelper(testCase.code)
    if (scopedTree === undefined) {
      return
    }
    testCase.targets.forEach(target => {
      actuals.push(
        result(
          testCase,
          target,
          getAllOccurrencesInScope(target.name, target.row, target.col, scopedTree)
        )
      )
    })
  })
  expect(actuals).toMatchSnapshot()
})

test('Going to definitions with arrow function scopes', () => {
  const actuals: any = []
  arrowFunctionScopeTests.forEach(testCase => {
    const program = testParseHelper(testCase.code)
    if (program === undefined) {
      return
    }
    const scopedTree = scopeVariables(program)
    testCase.targets.forEach(target => {
      actuals.push(
        result(testCase, target, lookupDefinition(target.name, target.row, target.col, scopedTree))
      )
    })
  })
  expect(actuals).toMatchSnapshot()
})

test('Scoped based refactoring with arrow function scopes', () => {
  const actuals: any = []
  arrowFunctionScopeTests.forEach(testCase => {
    const scopedTree = testParseHelper(testCase.code)
    if (scopedTree === undefined) {
      return
    }
    testCase.targets.forEach(target => {
      actuals.push(
        result(
          testCase,
          target,
          getAllOccurrencesInScope(target.name, target.row, target.col, scopedTree)
        )
      )
    })
  })
  expect(actuals).toMatchSnapshot()
})

test('Going to definitions of with conditionals and loops', () => {
  const actuals: any = []
  conditionalsLoopsTests.forEach(testCase => {
    const program = testParseHelper(testCase.code)
    if (program === undefined) {
      return
    }
    const scopedTree = scopeVariables(program)
    testCase.targets.forEach(target => {
      actuals.push(
        result(testCase, target, lookupDefinition(target.name, target.row, target.col, scopedTree))
      )
    })
  })
  expect(actuals).toMatchSnapshot()
})

test('Scoped based refactoring with conditionals and loops', () => {
  const actuals: any = []
  conditionalsLoopsTests.forEach(testCase => {
    const scopedTree = testParseHelper(testCase.code)
    if (scopedTree === undefined) {
      return
    }
    testCase.targets.forEach(target => {
      actuals.push(
        result(
          testCase,
          target,
          getAllOccurrencesInScope(target.name, target.row, target.col, scopedTree)
        )
      )
    })
  })
  expect(actuals).toMatchSnapshot()
})
