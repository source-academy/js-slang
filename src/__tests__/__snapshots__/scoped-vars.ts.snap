// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Going to definitions of ordinary variable definitions 1`] = `
Array [
  Object {
    "code": "let target = 1;
target = 2;
{
const target = 3;
target + 4;
}",
    "label": "Test 1",
    "target": Target {
      "col": 3,
      "name": "target",
      "row": 5,
    },
    "value": Object {
      "isDeclaration": true,
      "loc": SourceLocation {
        "end": Position {
          "column": 12,
          "line": 4,
        },
        "start": Position {
          "column": 6,
          "line": 4,
        },
      },
      "name": "target",
      "type": "DefinitionNode",
    },
  },
  Object {
    "code": "let target = 1;
target = 2;
target + 99;",
    "label": "Test 2",
    "target": Target {
      "col": 3,
      "name": "target",
      "row": 3,
    },
    "value": Object {
      "isDeclaration": false,
      "loc": SourceLocation {
        "end": Position {
          "column": 6,
          "line": 2,
        },
        "start": Position {
          "column": 0,
          "line": 2,
        },
      },
      "name": "target",
      "type": "DefinitionNode",
    },
  },
  Object {
    "code": "let target = 1;
target = 2;
target + 99;",
    "label": "Test 2",
    "target": Target {
      "col": 3,
      "name": "MISSING",
      "row": 3,
    },
    "value": undefined,
  },
]
`;

exports[`Going to definitions of with conditionals and loops 1`] = `
Array [
  Object {
    "code": "let target = 1;
if (target < 2) {
target = 2 + 3;
} else if (target > 2) {
target = \\"nooooooo\\";
} else {
target = 99999;
}",
    "label": "Test 1",
    "target": Target {
      "col": 7,
      "name": "target",
      "row": 5,
    },
    "value": Object {
      "isDeclaration": false,
      "loc": SourceLocation {
        "end": Position {
          "column": 6,
          "line": 5,
        },
        "start": Position {
          "column": 0,
          "line": 5,
        },
      },
      "name": "target",
      "type": "DefinitionNode",
    },
  },
]
`;

exports[`Going to definitions of with function scopes 1`] = `
Array [
  Object {
    "code": "let target = 1;
target = 2;
function test(target) {
const x = 3;
target + 4;
function y() {
target * 99;
}
}",
    "label": "Test 1",
    "target": Target {
      "col": 7,
      "name": "target",
      "row": 7,
    },
    "value": Object {
      "isDeclaration": true,
      "loc": SourceLocation {
        "end": Position {
          "column": 20,
          "line": 3,
        },
        "start": Position {
          "column": 14,
          "line": 3,
        },
      },
      "name": "target",
      "type": "DefinitionNode",
    },
  },
  Object {
    "code": "let target = 1;
target = 2;
function test(target) {
const x = 3;
target + 4;
function y() {
target * 99;
}
}",
    "label": "Test 1",
    "target": Target {
      "col": 6,
      "name": "target",
      "row": 5,
    },
    "value": Object {
      "isDeclaration": true,
      "loc": SourceLocation {
        "end": Position {
          "column": 20,
          "line": 3,
        },
        "start": Position {
          "column": 14,
          "line": 3,
        },
      },
      "name": "target",
      "type": "DefinitionNode",
    },
  },
]
`;

exports[`Going to definitions with arrow function scopes 1`] = `
Array [
  Object {
    "code": "let target = 1;
function test(target) {
const arrowFn = target => target + 1;
const x = target * 777777;
const nestedArrowFn = (y) => (target) => 1 + target;
}",
    "label": "Test 1",
    "target": Target {
      "col": 53,
      "name": "target",
      "row": 5,
    },
    "value": Object {
      "isDeclaration": true,
      "loc": SourceLocation {
        "end": Position {
          "column": 20,
          "line": 2,
        },
        "start": Position {
          "column": 14,
          "line": 2,
        },
      },
      "name": "target",
      "type": "DefinitionNode",
    },
  },
]
`;

exports[`Scoped based refactoring of ordinary variable definitions 1`] = `
Array [
  Object {
    "code": "let target = 1;
target = 2;
{
const target = 3;
target + 4;
}",
    "label": "Test 1",
    "target": Target {
      "col": 3,
      "name": "target",
      "row": 5,
    },
    "value": Array [
      SourceLocation {
        "end": Position {
          "column": 12,
          "line": 4,
        },
        "start": Position {
          "column": 6,
          "line": 4,
        },
      },
      SourceLocation {
        "end": Position {
          "column": 6,
          "line": 5,
        },
        "start": Position {
          "column": 0,
          "line": 5,
        },
      },
    ],
  },
  Object {
    "code": "let target = 1;
target = 2;
target + 99;",
    "label": "Test 2",
    "target": Target {
      "col": 3,
      "name": "target",
      "row": 3,
    },
    "value": Array [
      SourceLocation {
        "end": Position {
          "column": 10,
          "line": 1,
        },
        "start": Position {
          "column": 4,
          "line": 1,
        },
      },
      SourceLocation {
        "end": Position {
          "column": 6,
          "line": 2,
        },
        "start": Position {
          "column": 0,
          "line": 2,
        },
      },
      SourceLocation {
        "end": Position {
          "column": 6,
          "line": 3,
        },
        "start": Position {
          "column": 0,
          "line": 3,
        },
      },
    ],
  },
  Object {
    "code": "let target = 1;
target = 2;
target + 99;",
    "label": "Test 2",
    "target": Target {
      "col": 3,
      "name": "MISSING",
      "row": 3,
    },
    "value": Array [],
  },
]
`;

exports[`Scoped based refactoring with arrow function scopes 1`] = `
Array [
  Object {
    "code": "let target = 1;
function test(target) {
const arrowFn = target => target + 1;
const x = target * 777777;
const nestedArrowFn = (y) => (target) => 1 + target;
}",
    "label": "Test 1",
    "target": Target {
      "col": 53,
      "name": "target",
      "row": 5,
    },
    "value": Array [
      SourceLocation {
        "end": Position {
          "column": 20,
          "line": 2,
        },
        "start": Position {
          "column": 14,
          "line": 2,
        },
      },
      SourceLocation {
        "end": Position {
          "column": 16,
          "line": 4,
        },
        "start": Position {
          "column": 10,
          "line": 4,
        },
      },
    ],
  },
]
`;

exports[`Scoped based refactoring with conditionals and loops 1`] = `
Array [
  Object {
    "code": "let target = 1;
if (target < 2) {
target = 2 + 3;
} else if (target > 2) {
target = \\"nooooooo\\";
} else {
target = 99999;
}",
    "label": "Test 1",
    "target": Target {
      "col": 7,
      "name": "target",
      "row": 5,
    },
    "value": Array [
      SourceLocation {
        "end": Position {
          "column": 10,
          "line": 1,
        },
        "start": Position {
          "column": 4,
          "line": 1,
        },
      },
      SourceLocation {
        "end": Position {
          "column": 10,
          "line": 2,
        },
        "start": Position {
          "column": 4,
          "line": 2,
        },
      },
      SourceLocation {
        "end": Position {
          "column": 17,
          "line": 4,
        },
        "start": Position {
          "column": 11,
          "line": 4,
        },
      },
      SourceLocation {
        "end": Position {
          "column": 6,
          "line": 3,
        },
        "start": Position {
          "column": 0,
          "line": 3,
        },
      },
      SourceLocation {
        "end": Position {
          "column": 6,
          "line": 5,
        },
        "start": Position {
          "column": 0,
          "line": 5,
        },
      },
      SourceLocation {
        "end": Position {
          "column": 6,
          "line": 7,
        },
        "start": Position {
          "column": 0,
          "line": 7,
        },
      },
    ],
  },
]
`;

exports[`Scoped based refactoring with function scopes 1`] = `
Array [
  Object {
    "code": "let target = 1;
target = 2;
function test(target) {
const x = 3;
target + 4;
function y() {
target * 99;
}
}",
    "label": "Test 1",
    "target": Target {
      "col": 7,
      "name": "target",
      "row": 7,
    },
    "value": Array [
      SourceLocation {
        "end": Position {
          "column": 20,
          "line": 3,
        },
        "start": Position {
          "column": 14,
          "line": 3,
        },
      },
      SourceLocation {
        "end": Position {
          "column": 6,
          "line": 5,
        },
        "start": Position {
          "column": 0,
          "line": 5,
        },
      },
      SourceLocation {
        "end": Position {
          "column": 6,
          "line": 7,
        },
        "start": Position {
          "column": 0,
          "line": 7,
        },
      },
    ],
  },
  Object {
    "code": "let target = 1;
target = 2;
function test(target) {
const x = 3;
target + 4;
function y() {
target * 99;
}
}",
    "label": "Test 1",
    "target": Target {
      "col": 6,
      "name": "target",
      "row": 5,
    },
    "value": Array [
      SourceLocation {
        "end": Position {
          "column": 20,
          "line": 3,
        },
        "start": Position {
          "column": 14,
          "line": 3,
        },
      },
      SourceLocation {
        "end": Position {
          "column": 6,
          "line": 5,
        },
        "start": Position {
          "column": 0,
          "line": 5,
        },
      },
      SourceLocation {
        "end": Position {
          "column": 6,
          "line": 7,
        },
        "start": Position {
          "column": 0,
          "line": 7,
        },
      },
    ],
  },
]
`;
